src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
         end
     end
 
-    pp = similar(p)
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    if fun === Base.invpermute!! 
+        pop!(cp)
+        reverse!(cp)
+        push!(cp, 0)
+    end
 
     for (i, col) in enumerate(eachcol(df))
         if !(i in toskip)
-            copyto!(pp, p)
-            fun(col, pp)
+            _cycle_permute!(col, cp)
         end
     end
     return df
 end
 
+# convert a classical permutation to cycle notation, 
+# destroying the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    Base.require_one_based_indexing(p)
+    out = similar(p, 3length(p)÷2)
+    out_len = 0
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start+1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
+        k = p[start]
+        count -= 1
+        k == start && continue
+        out[out_len += 1] = k
```

Comment:
can you pleaes split here and below `out[out_len += 1] = k` into two lines?

Reply:
done